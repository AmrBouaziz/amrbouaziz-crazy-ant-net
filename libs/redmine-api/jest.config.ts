/* eslint-disable */
export default {
  displayName: 'redmine-api',
  preset: '../../jest.preset.js',
  transform: {
    '^.+\\.[tj]s$': ['ts-jest', { tsconfig: '<rootDir>/tsconfig.spec.json' }],
  },
  moduleFileExtensions: ['ts', 'js', 'html'],
  coverageDirectory: '../../coverage/libs/redmine-api',
};
