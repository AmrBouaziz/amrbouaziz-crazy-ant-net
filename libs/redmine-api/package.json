{
  "name": "@crazy-ant/redmine-api",
  "version": "0.0.1",
  "type": "commonjs"
}
