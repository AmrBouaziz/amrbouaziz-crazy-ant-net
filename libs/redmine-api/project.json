{
  "name": "redmine-api",
  "$schema": "../../node_modules/nx/schemas/project-schema.json",
  "sourceRoot": "libs/redmine-api/src",
  "projectType": "library",
  "targets": {
    "build": {
      "executor": "@nx/vite:build",
      "outputs": ["{options.outputPath}"],
      "options": {
        "outputPath": "dist/libs/redmine-api"
      }
    },
    "lint": {
      "executor": "@nx/linter:eslint",
      "outputs": ["{options.outputFile}"],
      "options": {
        "lintFilePatterns": ["libs/redmine-api/**/*.ts"]
      }
    },
    "test": {
      "executor": "@nx/jest:jest",
      "outputs": ["{workspaceRoot}/coverage/{projectRoot}"],
      "options": {
        "jestConfig": "libs/redmine-api/jest.config.ts",
        "passWithNoTests": true
      },
      "configurations": {
        "ci": {
          "ci": true,
          "codeCoverage": true
        }
      }
    }
  },
  "tags": []
}
