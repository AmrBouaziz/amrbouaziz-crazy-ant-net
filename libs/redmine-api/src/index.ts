export * from './lib/redmine-api';
