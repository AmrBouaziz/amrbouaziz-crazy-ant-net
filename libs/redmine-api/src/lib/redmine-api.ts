import { Axios } from 'axios-observable';
import { forkJoin, map, Observable, tap } from 'rxjs';
import { XMLBuilder } from 'fast-xml-parser';
import { AxiosResponse } from 'axios';

export  const  ACTIVITY_MAP = {
  DEV: 9,
  MAINTENANCE: 10,
  SUPPORT: 15,
}

export type Activity = keyof typeof ACTIVITY_MAP;
export class RedmineApi {



  private readonly _xmlBuilder = new XMLBuilder({
    ignoreAttributes: false,
    format: true,
  });
  constructor(
    private readonly _redmineBaseUrl: string,
    private readonly _redmineAccessKey: string,
  ) {}

  buildXmlBody(date: string, issue: number, comment: string = 'No Comment', activity: Activity= 'DEV') {
    return this._xmlBuilder.build({
      time_entry: {
        issue_id: issue,
        activity_id: ACTIVITY_MAP[activity], // DEV
        hours: 8.0,
        comments: comment,
        spent_on: date.replace(/\//g, '-'),
      },
    });
  }

  markDayAsWorked<
    T extends string | string[],
    ReturnType = T extends string[]
      ? Array<AxiosResponse<string>>
      : T extends string
      ? AxiosResponse<string>
      : never,
  >(date: T, issue: number, comment?: string, activity?: Activity): Observable<ReturnType> {
    let dates: string[];
    let singleValue = false;
    if (typeof date === 'string') {
      singleValue = true;
      dates = [date];
    } else if (Array.isArray(date)) {
      dates = date;
    } else {
      throw new Error(
        `expected date to be string | string[] but got ${typeof date}`,
      );
    }

    return forkJoin(
      dates.map(date =>
        Axios.post(
          this._redmineBaseUrl + '/time_entries.json',
          this.buildXmlBody(date, issue, comment, activity),
          {
            headers: {
              'Content-Type': 'application/xml',
              'X-Redmine-API-Key': this._redmineAccessKey
            }
          }
        )
      )
    ).pipe(
      // @ts-ignore TODO: No any please
      map((value: any) => {
        if (singleValue) {
          return value[0] as ReturnType;
        }
        return value as ReturnType;
      }),
      tap(console.log)
    );
  }
}
