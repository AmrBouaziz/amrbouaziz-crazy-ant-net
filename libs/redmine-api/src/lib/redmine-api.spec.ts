import { RedmineApi } from './redmine-api';
import { lastValueFrom, of } from 'rxjs';
import { XMLBuilder } from 'fast-xml-parser';
import { Axios } from 'axios-observable';

const redmineBaseUrl = 'https://redmine.squeezer-software.com';
const redmineAccessKey = 'a6120097f9989fabe5947dccd1f5879a8e8a4e3a'; // TODO : TOP SECRET. should handle properly

describe(RedmineApi.name, () => {
  let redmineApi: RedmineApi;
  beforeEach(() => {
    // Note: this is required while in jest test because jest-dom provides the xhr API found in the browser.
    //  usually this is not a bad thing, in fact, this leads me to redo the tests for this lib (using snapshots maybe)
    //  however, I'm calling the real server, and xhr neads to make a preflight request to check if the request should be allowed
    //  see more at https://github.com/axios/axios/issues/1180
    Axios.defaults.adapter = 'http';

    redmineApi = new RedmineApi(
      redmineBaseUrl,
      redmineAccessKey,
    );
  });

  describe(RedmineApi.prototype.buildXmlBody.name, () => {
    it('should build xml body', () => {
      expect(redmineApi.buildXmlBody('2021/12/31', issue, comment)).toMatchSnapshot();
    });
  });

  describe(RedmineApi.prototype.markDayAsWorked.name, () => {
    it('should call the correct endpoint', async () => {
      // Given
      jest.spyOn(Axios, 'post').mockReturnValue(
        of({
          data: `<?xml version="1.0" encoding="UTF-8"?>
<time_entry>
    <id>51419</id>
    <project id="205" name="[Ouimind-FR] Mission Clevermind - LCL"/>
    <issue id="26151"/>
    <user id="43" name="Amr BOUAZIZ"/>
    <activity id="9" name="Development"/>
    <hours>1.0</hours>
    <comments>LCL</comments>
    <spent_on>2021-12-31</spent_on>
    <created_on>2023-03-11T14:21:13Z</created_on>
    <updated_on>2023-03-11T14:21:13Z</updated_on>
</time_entry>`,
        }) as any, // TODO : find a better type
      );

      const date = '2021/12/31';
      const requestBody = `<time_entry>
  <issue_id>26151</issue_id>
  <activity_id>9</activity_id>
  <hours>8</hours>
  <comments>LCL</comments>
  <spent_on>2021-12-31</spent_on>
</time_entry>
`;
      // WHEN

      const response = await lastValueFrom(redmineApi.markDayAsWorked(date));
      expect(Axios.post).toBeCalledTimes(1);
      expect(Axios.post).toBeCalledWith(
        redmineBaseUrl + '/time_entries.json',
        requestBody,
        {
          headers: {
            'Content-Type': 'application/xml',
            'X-Redmine-API-Key': redmineAccessKey,
          },
        },
      );
      expect(response).toBeDefined();
      expect(response).not.toBeNull();
      expect(Array.isArray(response)).toBeFalsy();
    });
    it('should call the correct endpoint twice', async () => {
      // Given
      jest.spyOn(Axios, 'post').mockReturnValue(
        of({
          data: `<?xml version="1.0" encoding="UTF-8"?>
<time_entry>
    <id>51419</id>
    <project id="205" name="[Ouimind-FR] Mission Clevermind - LCL"/>
    <issue id="26151"/>
    <user id="43" name="Amr BOUAZIZ"/>
    <activity id="9" name="Development"/>
    <hours>1.0</hours>
    <comments>LCL</comments>
    <spent_on>2021-12-31</spent_on>
    <created_on>2023-03-11T14:21:13Z</created_on>
    <updated_on>2023-03-11T14:21:13Z</updated_on>
</time_entry>`,
        }) as any, // TODO : find a better type
      );

      const date = '2021/12/31';
      const requestBody = `<time_entry>
  <issue_id>26151</issue_id>
  <activity_id>9</activity_id>
  <hours>8</hours>
  <comments>LCL</comments>
  <spent_on>2021-12-31</spent_on>
</time_entry>
`;
      // WHEN

      const response = await lastValueFrom(redmineApi.markDayAsWorked([date, date]));
      expect(Axios.post).toBeCalledTimes(2);
      expect(Axios.post).toBeCalledWith(
        redmineBaseUrl + '/time_entries.json',
        requestBody,
        {
          headers: {
            'Content-Type': 'application/xml',
            'X-Redmine-API-Key': redmineAccessKey,
          },
        },
      );
      expect(response).toBeDefined();
      expect(response).not.toBeNull();
      expect(Array.isArray(response)).toBeTruthy();
      expect(response.length).toEqual(2)
    });
  });
});
