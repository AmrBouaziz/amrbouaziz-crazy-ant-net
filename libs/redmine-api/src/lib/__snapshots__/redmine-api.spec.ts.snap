// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RedmineApi buildXmlBody should build xml body 1`] = `
"<time_entry>
  <issue_id>26151</issue_id>
  <activity_id>9</activity_id>
  <hours>8</hours>
  <comments>LCL</comments>
  <spent_on>2021-12-31</spent_on>
</time_entry>
"
`;
